File: go_patcher/src/patcher/services/patchService.go
package services
import (
	"fmt"
	libs "patcher/libs"
	utils "patcher/utils"
)

func GeneratePatch(params []string) bool{
	res := libs.Generate(params[0], params[1], params[2])
	if (!res){
		fmt.Println("Error Generating Patch")
	}
	return res
}

func ApplyPatch(params []string) bool{
	res := libs.Apply(params[0], params[1], params[2])
	if (!res) {
		fmt.Println("Error Applying Patch")
	}
	return res
}

func GeneratePatchWithCrc32(params []string) bool{
	res := GeneratePatch(params[:3])
	if (!res) {
		return false
	}
	new := params[1]
	crc32, err := libs.ComputeCrc32(new)
	if (err != nil){
		fmt.Println("Error Generating crc32")
		return false
	}
	crc32FileName := params[3]
	res = utils.WriteStringToFile(crc32FileName, crc32)
	if (!res) {
		fmt.Println("Error Writing crc32 file")
		return false
	}
	return true
}

func ApplyPatchWithCrc32(params []string) bool{
	res := ApplyPatch(params[:3])
	if (!res) {
		return false
	}
	patched := params[2]
	crc32Patched, err := libs.ComputeCrc32(patched)
	if (err != nil){
		fmt.Println("Error Generating crc32")
		return false
	}
	crc32FileName := params[3]
	crc32NewFile, resr :=utils.ReadStringFromFile(crc32FileName) 
	if (!resr){
		fmt.Println("Error reading crc32 file")
		return false
	}
	if (crc32Patched == crc32NewFile){
		//fmt.Println("Success")
		return true
	}
	fmt.Println("crc32 not equals")
	return false
}

func PrintCrc32(params []string) bool{
	crc32, err := libs.ComputeCrc32(params[0])
	if (err != nil){
		fmt.Println("Error Computing crc32")
		return false
	}
	fmt.Println(crc32)
	return true
}

File: patcher/services/patchService.py
import patchLib, utils

def generatePatch(old, new, patch):
    res = patchLib.generate(old, new, patch)
    if(res != 0):
        print("Error Generating Patch")
        return False
    return True

def applyPatch(old, patch, patched):
    res = patchLib.apply(old, patch, patched)
    if(res != 0):
        print("Error Applying Patch")
        return False
    return True
    
def generatePatchWithCrc32(old, new, patch, crc32File):
    patchRes = generatePatch(old, new, patch)
    if (not patchRes):
        return False
    crcRes, crc32 = patchLib.computeCrc32(new)
    if (not crcRes):
        print("Error generating crc32")
        return False
    writeCrcFileRes = utils.writeSmallFile(crc32File, crc32)
    if (not writeCrcFileRes):
        print("Error writing crc32 file")
        return False

def applyPatchWithCrc32(old, patch, patched, crc32File):
    patchRes = applyPatch(old, patch, patched)
    if (not patchRes):
        return False
    crcRes, crc32 = patchLib.computeCrc32(patched)
    if (not crcRes):
        print("Error generating crc32")
        return False
    readCrcFileRes, crcNewFile = utils.readSmallFile(crc32File)
    if(not readCrcFileRes):
        print("Error reading crc32 file")
        return False
    if(crcNewFile == crc32):
        print("Success")
        return True
    print("crc32 not equals")
    return False

    
